Regularisation Techniques to overcome Overfitting and Underfitting of the model:

1. Ridge : Penalty is added to the square of coefficients

2. Lasso : Penalty is added to the abs value of coefficients

3. Elastic Net : Combination of Ridge and Lasso

4. Dropout : Drop few features from the model. Randomly set some coefficients to zero

Read the dataset

In [2]:
from warnings import filterwarnings
filterwarnings("ignore")

In [1]:
path = r"D:\Machine-Learning\repository\Data_Processing\Cars93.csv"

In [3]:
import pandas as pd
df = pd.read_csv(path,na_values=["","NA"],keep_default_na=False)
df.head()

id Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0   1        Acura  Integra    Small       12.9   15.9       18.8        25   
1   2        Acura   Legend  Midsize       29.2   33.9       38.7        18   
2   3         Audi       90  Compact       25.9   29.1       32.3        20   
3   4         Audi      100  Midsize       30.8   37.7       44.6        19   
4   5          BMW     535i  Midsize       23.7   30.0       36.2        22   

   MPG.highway             AirBags  ... Passengers Length  Wheelbase  Width  \
0           31                None  ...          5    177        102     68   
1           25  Driver & Passenger  ...          5    195        115     71   
2           26         Driver only  ...          5    180        102     67   
3           26                 NaN  ...          6    193        106     70   
4           30         Driver only  ...          4    186        109     69   

   Turn.circle  Rear.seat.room Luggage.room  Weight   Origin           Make  
0           37            26.5         11.0    2705  non-USA  Acura Integra  
1           38            30.0         15.0    3560  non-USA   Acura Legend  
2           37            28.0         14.0    3375  non-USA        Audi 90  
3           37            31.0         17.0    3405  non-USA       Audi 100  
4           39            27.0         13.0    3640  non-USA       BMW 535i  

[5 rows x 28 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93 entries, 0 to 92
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id                  93 non-null     int64  
 1   Manufacturer        93 non-null     object 
 2   Model               93 non-null     object 
 3   Type                93 non-null     object 
 4   Min.Price           93 non-null     float64
 5   Price               93 non-null     float64
 6   Max.Price           93 non-null     float64
 7   MPG.city            93 non-null     int64  
 8   MPG.highway         93 non-null     int64  
 9   AirBags             89 non-null     object 
 10  DriveTrain          93 non-null     object 
 11  Cylinders           93 non-null     object 
 12  EngineSize          93 non-null     float64
 13  Horsepower          93 non-null     int64  
 14  RPM                 93 non-null     int64  
 15  Rev.per.mile        93 non-null     int64  
 16  Man.trans.

In [5]:
df.isna().sum()

id                     0
Manufacturer           0
Model                  0
Type                   0
Min.Price              0
Price                  0
Max.Price              0
MPG.city               0
MPG.highway            0
AirBags                4
DriveTrain             0
Cylinders              0
EngineSize             0
Horsepower             0
RPM                    0
Rev.per.mile           0
Man.trans.avail        0
Fuel.tank.capacity     0
Passengers             0
Length                 0
Wheelbase              0
Width                  0
Turn.circle            0
Rear.seat.room         2
Luggage.room          11
Weight                 0
Origin                 0
Make                   0
dtype: int64

In [6]:
df.duplicated().sum()

0

Separate X and Y features

In [7]:
X = df.drop(columns=["id","Weight"])
Y = df[["Weight"]]

In [8]:
cat = list(X.columns[X.dtypes=="object"])
con = list(X.columns[X.dtypes!="object"])

Feature Engineering: Data Preprocessing and Data Cleaning

In [9]:
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder,StandardScaler
from sklearn.compose import ColumnTransformer

In [10]:
cat_pipe = make_pipeline(
    SimpleImputer(strategy="most_frequent"),
    OneHotEncoder(handle_unknown='ignore',sparse_output=False)
)

In [11]:
con_pipe = make_pipeline(
    SimpleImputer(strategy="median"),
    StandardScaler()
)

In [12]:
pre = ColumnTransformer([
    ("cat",cat_pipe,cat),
    ("con",con_pipe,con)
]).set_output(transform="pandas")

In [13]:
X_pre = pre.fit_transform(X)
X_pre.head()

cat__Manufacturer_Acura  cat__Manufacturer_Audi  cat__Manufacturer_BMW  \
0                      1.0                     0.0                    0.0   
1                      1.0                     0.0                    0.0   
2                      0.0                     1.0                    0.0   
3                      0.0                     1.0                    0.0   
4                      0.0                     0.0                    1.0   

   cat__Manufacturer_Buick  cat__Manufacturer_Cadillac  \
0                      0.0                         0.0   
1                      0.0                         0.0   
2                      0.0                         0.0   
3                      0.0                         0.0   
4                      0.0                         0.0   

   cat__Manufacturer_Chevrolet  cat__Manufacturer_Chrylser  \
0                          0.0                         0.0   
1                          0.0                         0.0   
2                          0.0                         0.0   
3                          0.0                         0.0   
4                          0.0                         0.0   

   cat__Manufacturer_Chrysler  cat__Manufacturer_Dodge  \
0                         0.0                      0.0   
1                         0.0                      0.0   
2                         0.0                      0.0   
3                         0.0                      0.0   
4                         0.0                      0.0   

   cat__Manufacturer_Eagle  ...  con__RPM  con__Rev.per.mile  \
0                      0.0  ...  1.717489           1.129530   
1                      0.0  ...  0.369586           0.005661   
2                      0.0  ...  0.369586          -0.105713   
3                      0.0  ...  0.369586           0.410659   
4                      0.0  ...  0.706562           0.430909   

   con__Fuel.tank.capacity  con__Passengers  con__Length  con__Wheelbase  \
0                -1.062184        -0.083243    -0.427186       -0.286932   
1                 0.409445        -0.083243     0.812171        1.629649   
2                 0.072197        -0.083243    -0.220626       -0.286932   
3                 1.359872         0.884457     0.674465        0.302785   
4                 1.359872        -1.050944     0.192493        0.745073   

   con__Width  con__Turn.circle  con__Rear.seat.room  con__Luggage.room  
0   -0.366184         -0.610436            -0.449727          -1.037572  
1    0.431983         -0.298510             0.740403           0.391972  
2   -0.632239         -0.610436             0.060329           0.034586  
3    0.165927         -0.610436             1.080441           1.106743  
4   -0.100128          0.013416            -0.279708          -0.322800  

[5 rows x 257 columns]

Apply train test split

In [14]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest = train_test_split(X_pre,Y,train_size=0.8,random_state=42)

In [15]:

xtrain.head()

cat__Manufacturer_Acura  cat__Manufacturer_Audi  cat__Manufacturer_BMW  \
65                      0.0                     0.0                    0.0   
15                      0.0                     0.0                    0.0   
68                      0.0                     0.0                    0.0   
78                      0.0                     0.0                    0.0   
30                      0.0                     0.0                    0.0   

    cat__Manufacturer_Buick  cat__Manufacturer_Cadillac  \
65                      0.0                         0.0   
15                      0.0                         0.0   
68                      0.0                         0.0   
78                      0.0                         0.0   
30                      0.0                         0.0   

    cat__Manufacturer_Chevrolet  cat__Manufacturer_Chrylser  \
65                          0.0                         0.0   
15                          1.0                         0.0   
68                          0.0                         0.0   
78                          0.0                         0.0   
30                          0.0                         0.0   

    cat__Manufacturer_Chrysler  cat__Manufacturer_Dodge  \
65                         0.0                      0.0   
15                         0.0                      0.0   
68                         0.0                      0.0   
78                         0.0                      0.0   
30                         0.0                      0.0   

    cat__Manufacturer_Eagle  ...  con__RPM  con__Rev.per.mile  \
65                      0.0  ... -0.809828          -0.541086   
15                      0.0  ... -0.809828          -1.300456   
68                      0.0  ... -0.135877           0.471409   
78                      0.0  ... -0.472853          -0.379087   
30                      0.0  ... -0.472853           1.656027   

    con__Fuel.tank.capacity  con__Passengers  con__Length  con__Wheelbase  \
65                 1.022624         1.852158     0.467905        1.187361   
15                 1.022624         1.852158    -0.358333        0.892502   
68                -0.050439        -0.083243     0.467905        0.155356   
78                -1.184820        -0.083243    -0.496039       -0.286932   
30                -2.043270        -1.050944    -2.905899       -2.056085   

    con__Width  con__Turn.circle  con__Rear.seat.room  con__Luggage.room  
65    1.230149          0.637268            -0.279708           0.034586  
15    1.230149          1.573046             0.910422           0.034586  
68    0.165927          0.949194             0.060329           0.749357  
78   -0.366184          0.325342            -0.449727          -0.680186  
30   -1.696461         -1.858140            -0.619745          -0.680186  

[5 rows x 257 columns]

In [16]:
xtest.head()

cat__Manufacturer_Acura  cat__Manufacturer_Audi  cat__Manufacturer_BMW  \
40                      0.0                     0.0                    0.0   
22                      0.0                     0.0                    0.0   
55                      0.0                     0.0                    0.0   
72                      0.0                     0.0                    0.0   
0                       1.0                     0.0                    0.0   

    cat__Manufacturer_Buick  cat__Manufacturer_Cadillac  \
40                      0.0                         0.0   
22                      0.0                         0.0   
55                      0.0                         0.0   
72                      0.0                         0.0   
0                       0.0                         0.0   

    cat__Manufacturer_Chevrolet  cat__Manufacturer_Chrylser  \
40                          0.0                         0.0   
22                          0.0                         0.0   
55                          0.0                         0.0   
72                          0.0                         0.0   
0                           0.0                         0.0   

    cat__Manufacturer_Chrysler  cat__Manufacturer_Dodge  \
40                         0.0                      0.0   
22                         0.0                      1.0   
55                         0.0                      0.0   
72                         0.0                      0.0   
0                          0.0                      0.0   

    cat__Manufacturer_Eagle  ...  con__RPM  con__Rev.per.mile  \
40                      0.0  ...  0.875050           1.058655   
22                      0.0  ...  1.212025           1.929401   
55                      0.0  ... -0.472853          -0.186713   
72                      0.0  ...  0.538074           1.615527   
0                       0.0  ...  1.717489           1.129530   

    con__Fuel.tank.capacity  con__Passengers  con__Length  con__Wheelbase  \
40                -0.234393        -1.050944    -0.564892       -0.581791   
22                -1.062184        -0.083243    -0.633745       -0.876650   
55                 0.899988         1.852158     0.467905        0.892502   
72                -1.062184        -1.050944    -0.427186       -0.729220   
0                 -1.062184        -0.083243    -0.427186       -0.286932   

    con__Width  con__Turn.circle  con__Rear.seat.room  con__Luggage.room  
40    0.165927          0.013416            -1.469838          -2.109729  
22   -0.898295         -2.170066            -0.449727          -1.037572  
55    0.698038          0.013416            -0.109689           0.034586  
72   -0.898295         -1.234288            -0.789764           1.106743  
0    -0.366184         -0.610436            -0.449727          -1.037572  

[5 rows x 257 columns]

In [17]:
ytrain.head()

Weight
65    4100
15    3715
68    2890
78    2495
30    1845

In [18]:
ytest.head()

Weight
40    2865
22    2270
55    3735
72    2350
0     2705

Model Buidling

a) Linear Regression

In [19]:
from sklearn.linear_model import LinearRegression,Ridge
model = LinearRegression()
model.fit(xtrain,ytrain)

LinearRegression()

In [20]:
model.score(xtrain,ytrain)

1.0

In [21]:
model.score(xtest,ytest)

0.9216585920409737

In [22]:
model2 = Ridge(alpha=1)
model2.fit(xtrain,ytrain)

Ridge(alpha=1)

In [23]:
model2.score(xtrain,ytrain)

0.9980089509696602

In [24]:
model2.score(xtest,ytest)

0.9265719260997733

Hyperparameter tuning

In [25]:
import numpy as np
np.arange(start=0.1,stop=100,step=0.1)

array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
        7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
        8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
       10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
       11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
       12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
       13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
       14.4, 14.5, 14.6, 14.7, 14.8, 14.9, 15. , 15

In [26]:
params = {
    "alpha":np.arange(start=0.1,stop=100,step=0.1)
}

In [27]:
params.items()

dict_items([('alpha', array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
        7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
        8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
       10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
       11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
       12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
       13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
       14.4, 14.5, 14.6, 14.7

GridSearchCV

RandomisedSearchCV

In [28]:
from sklearn.model_selection import GridSearchCV
base_model = Ridge()
gscv = GridSearchCV(estimator=base_model,param_grid=params,scoring="r2")
gscv.fit(xtrain,ytrain)

GridSearchCV(estimator=Ridge(),
             param_grid={'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5...
       92.5, 92.6, 92.7, 92.8, 92.9, 93. , 93.1, 93.2, 93.3, 93.4, 93.5,
       93.6, 93.7, 93.8, 93.9, 94. , 94.1, 94.2, 94.3, 94.4, 94.5, 94.6,
       94.7, 94.8, 94.9, 95. , 95.1, 95.2, 95.3, 95.4, 95.5, 95.6, 95.7,
       95.8, 95.9, 96. , 96.1, 96.2, 96.3, 96.4, 96.5, 96.6, 96.7, 96.8,
       96.9, 97. , 97.1, 97.2, 97.3, 97.4, 97.5, 97.6, 97.7, 97.8, 97.9,
       98. , 98.1, 98.2, 98.3, 98.4, 98.5, 98.6, 98.7, 98.8, 98.9, 99. ,
       99.1, 99.2, 99.3, 99.4, 99.5, 99.6, 99.7, 99.8, 99.9])},
             scoring='r2')

In [29]:
gscv.best_params_

{'alpha': 3.6}

In [30]:
gscv.best_score_

0.9345217537539913

In [31]:
best_ridge = gscv.best_estimator_
best_ridge

Ridge(alpha=3.6)

In [32]:
best_ridge.score(xtrain,ytrain)

0.9905907284418577

In [33]:
best_ridge.score(xtest,ytest)

0.9334034196243888

c) Lasso

In [34]:
from sklearn.linear_model import Lasso
model3 = Lasso(alpha=1)
model3.fit(xtrain,ytrain)

Lasso(alpha=1)

In [35]:
model3.score(xtrain,ytrain)

0.9927584856834794

In [36]:
model3.score(xtest,ytest)

0.9082145957223835

Hyperparamter tuning

In [37]:
params2 = {
    "alpha" : np.arange(start=0.1,stop=150,step=0.2)
}

In [38]:
base_model2 = Lasso()
gscv2 = GridSearchCV(estimator=base_model2,param_grid=params2,scoring="r2")
gscv2.fit(xtrain,ytrain)

GridSearchCV(estimator=Lasso(),
             param_grid={'alpha': array([1.000e-01, 3.000e-01, 5.000e-01, 7.000e-01, 9.000e-01, 1.100e+00,
       1.300e+00, 1.500e+00, 1.700e+00, 1.900e+00, 2.100e+00, 2.300e+00,
       2.500e+00, 2.700e+00, 2.900e+00, 3.100e+00, 3.300e+00, 3.500e+00,
       3.700e+00, 3.900e+00, 4.100e+00, 4.300e+00, 4.500e+00, 4.700e+00,
       4.900e+00, 5.100e+00, 5.300e+00, 5.500e+00, 5.700e+00, 5....
       1.429e+02, 1.431e+02, 1.433e+02, 1.435e+02, 1.437e+02, 1.439e+02,
       1.441e+02, 1.443e+02, 1.445e+02, 1.447e+02, 1.449e+02, 1.451e+02,
       1.453e+02, 1.455e+02, 1.457e+02, 1.459e+02, 1.461e+02, 1.463e+02,
       1.465e+02, 1.467e+02, 1.469e+02, 1.471e+02, 1.473e+02, 1.475e+02,
       1.477e+02, 1.479e+02, 1.481e+02, 1.483e+02, 1.485e+02, 1.487e+02,
       1.489e+02, 1.491e+02, 1.493e+02, 1.495e+02, 1.497e+02, 1.499e+02])},
             scoring='r2')

In [39]:
gscv2.best_params_

{'alpha': 0.9000000000000001}

In [40]:
gscv2.best_score_

0.9240155902452223

In [41]:
best_lasso = gscv2.best_estimator_
best_lasso

Lasso(alpha=0.9000000000000001)

In [42]:
best_lasso.score(xtrain,ytrain)

0.9938001607735617

In [43]:
best_lasso.score(xtest,ytest)

0.9075755315994071

Considering Ridge as best model for out of sample predictions,
 since it is providing good r2 score of 93.34%

Out of sample predictions

In [44]:
path2 = r"D:\Machine-Learning\repository\Data_Processing\sample_cars93.csv"

In [45]:
xnew = pd.read_csv(path2,na_values=["","NA"],keep_default_na=False)
xnew

Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0         Audi      100  Midsize       30.8   37.7       44.6        19   
1      Pontiac  Sunbird  Compact        9.4   11.1       12.8        23   
2    Chevrolet   Lumina  Midsize       13.4   15.9       18.4        21   
3        Mazda     RX-7   Sporty       32.5   32.5       32.5        17   
4   Volkswagen      Fox    Small        8.7    9.1        9.5        25   

   MPG.highway      AirBags DriveTrain  ... Fuel.tank.capacity  Passengers  \
0           26          NaN      Front  ...               15.0           6   
1           31         None      Front  ...               15.2           5   
2           29         None      Front  ...               16.5           6   
3           25  Driver only       Rear  ...               20.0           2   
4           33         None      Front  ...               12.4           4   

   Length  Wheelbase  Width Turn.circle  Rear.seat.room  Luggage.room  \
0     190        106     65          37            31.0          17.0   
1     181        101     66          39            25.0          13.0   
2     198        108     71          40            28.5          16.0   
3     169         96     69          37             NaN           NaN   
4     163         93     63          34            26.0          10.0   

    Origin              Make  
0  non-USA          Audi 100  
1      USA   Pontiac Sunbird  
2      USA  Chevrolet Lumina  
3  non-USA        Mazda RX-7  
4  non-USA    Volkswagen Fox  

[5 rows x 26 columns]

In [46]:
xnew_pre = pre.transform(xnew)
xnew_pre

cat__Manufacturer_Acura  cat__Manufacturer_Audi  cat__Manufacturer_BMW  \
0                      0.0                     1.0                    0.0   
1                      0.0                     0.0                    0.0   
2                      0.0                     0.0                    0.0   
3                      0.0                     0.0                    0.0   
4                      0.0                     0.0                    0.0   

   cat__Manufacturer_Buick  cat__Manufacturer_Cadillac  \
0                      0.0                         0.0   
1                      0.0                         0.0   
2                      0.0                         0.0   
3                      0.0                         0.0   
4                      0.0                         0.0   

   cat__Manufacturer_Chevrolet  cat__Manufacturer_Chrylser  \
0                          0.0                         0.0   
1                          0.0                         0.0   
2                          1.0                         0.0   
3                          0.0                         0.0   
4                          0.0                         0.0   

   cat__Manufacturer_Chrysler  cat__Manufacturer_Dodge  \
0                         0.0                      0.0   
1                         0.0                      0.0   
2                         0.0                      0.0   
3                         0.0                      0.0   
4                         0.0                      0.0   

   cat__Manufacturer_Eagle  ...  con__RPM  con__Rev.per.mile  \
0                      0.0  ...  0.369586           0.410659   
1                      0.0  ... -0.135877           0.673908   
2                      0.0  ... -0.135877           0.532158   
3                      0.0  ...  2.054464          -0.014589   
4                      0.0  ...  0.369586           0.441034   

   con__Fuel.tank.capacity  con__Passengers  con__Length  con__Wheelbase  \
0                -0.510323         0.884457     0.467905        0.302785   
1                -0.449005        -0.083243    -0.151773       -0.434362   
2                -0.050439         0.884457     1.018731        0.597644   
3                 1.022624        -2.986345    -0.978011       -1.171509   
4                -1.307455        -1.050944    -1.391130       -1.613797   

   con__Width  con__Turn.circle  con__Rear.seat.room  con__Luggage.room  
0   -1.164350         -0.610436             1.080441           1.106743  
1   -0.898295          0.013416            -0.959782          -0.322800  
2    0.431983          0.325342             0.230348           0.749357  
3   -0.100128         -0.610436            -0.109689           0.034586  
4   -1.696461         -1.546214            -0.619745          -1.394958  

[5 rows x 257 columns]

In [47]:
preds = best_ridge.predict(xnew_pre)
preds

array([3281.89034414, 2637.7917531 , 3109.37576488, 2963.31247568,
       2260.43947977])

Save the predictions back to xnew dataframe and save the results

In [48]:
xnew["Weight Predicted"] = preds
xnew = xnew.round(2)
xnew

Manufacturer    Model     Type  Min.Price  Price  Max.Price  MPG.city  \
0         Audi      100  Midsize       30.8   37.7       44.6        19   
1      Pontiac  Sunbird  Compact        9.4   11.1       12.8        23   
2    Chevrolet   Lumina  Midsize       13.4   15.9       18.4        21   
3        Mazda     RX-7   Sporty       32.5   32.5       32.5        17   
4   Volkswagen      Fox    Small        8.7    9.1        9.5        25   

   MPG.highway      AirBags DriveTrain  ... Passengers  Length  Wheelbase  \
0           26          NaN      Front  ...          6     190        106   
1           31         None      Front  ...          5     181        101   
2           29         None      Front  ...          6     198        108   
3           25  Driver only       Rear  ...          2     169         96   
4           33         None      Front  ...          4     163         93   

   Width  Turn.circle Rear.seat.room  Luggage.room   Origin              Make  \
0     65           37           31.0          17.0  non-USA          Audi 100   
1     66           39           25.0          13.0      USA   Pontiac Sunbird   
2     71           40           28.5          16.0      USA  Chevrolet Lumina   
3     69           37            NaN           NaN  non-USA        Mazda RX-7   
4     63           34           26.0          10.0  non-USA    Volkswagen Fox   

   Weight Predicted  
0           3281.89  
1           2637.79  
2           3109.38  
3           2963.31  
4           2260.44  

[5 rows x 27 columns]

In [49]:
xnew.to_csv("RidgeResults.csv",index=False)